internal/controller/management/serviceaccounts/serviceaccounts.go

```diff
@@ -0,0 +1,301 @@
+package serviceaccounts
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/kelseyhightower/envconfig"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	kubeerr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/logging"
+)
+
+const (
+	componentLabelKey    = "app.kubernetes.io/component"
+	controllerLabelValue = "controller"
+)
+
+type ReconcilerConfig struct {
+	KargoNamespace string `envconfig:"KARGO_NAMESPACE" required:"true"`
+}
+
+func ReconcilerConfigFromEnv() ReconcilerConfig {
+	cfg := ReconcilerConfig{}
+	envconfig.MustProcess("", &cfg)
+	return cfg
+}
+
+// reconciler reconciles ServiceAccount resources
+type reconciler struct {
+	cfg    ReconcilerConfig
+	client client.Client
+
+	getServiceAccountFn func(
+		context.Context,
+		types.NamespacedName,
+		client.Object,
+		...client.GetOption,
+	) error
+
+	updateServiceAccountFn func(
+		context.Context,
+		client.Object,
+		...client.UpdateOption,
+	) error
+
+	createRoleBindingFn func(
+		context.Context,
+		client.Object,
+		...client.CreateOption,
+	) error
+
+	updateRoleBindingFn func(
+		context.Context,
+		client.Object,
+		...client.UpdateOption,
+	) error
+
+	deleteRoleBindingFn func(
+		context.Context,
+		client.Object,
+		...client.DeleteOption,
+	) error
+
+	listProjectFn func(
+		context.Context,
+		client.ObjectList,
+		...client.ListOption,
+	) error
+
+	ensureControllerPermissionsFn func(context.Context, *corev1.ServiceAccount) error
+	removeControllerPermissionsFn func(context.Context, types.NamespacedName) error
```

Comment:
As discussed offline, these can go away as we are avoiding this approach to testing for new components.